gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -924,6 +932,218 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
+
+    String parentFolder = folderResource.getParentFolderName();
+    if (countOfChildren.containsKey(parentFolder)) {
+
+      // update the parent's count of children
+      countOfChildren.replace(parentFolder, countOfChildren.get(parentFolder) - 1);
+
+      // if the parent folder is now empty, append in the queue
+      if (countOfChildren.get(parentFolder) == 0) {
+        addFolderResourceInBlockingQueue(
+            new FolderInfo(
+                FolderInfo.createFolderInfoObject(folderResource.getBucket(), parentFolder)),
+            folderDeleteBlockingQueue,
+            innerExceptions);
+      }
+    }
+  }
+
+  /** Helper to create a callback for a particular deletion request for folder. */
+  private JsonBatchCallback<Void> getDeletionCallback(
+      final FolderInfo resourceId,
+      ConcurrentHashMap<String, Long> countOfChildren,
+      BlockingQueue<FolderInfo> folderDeleteBlockingQueue,
+      final KeySetView<IOException, Boolean> innerExceptions,
+      final BatchExecutor batchExecutor,
+      final int attempt) {
+    return new JsonBatchCallback<Void>() {
+      @Override
+      public void onSuccess(Void obj, HttpHeaders responseHeaders) {
+        logger.atFiner().log("Successfully deleted folder %s", resourceId.toString());
+        successfullDeletionOfFolderResource(
+            resourceId, countOfChildren, folderDeleteBlockingQueue, innerExceptions);
+      }
+
+      @Override
+      public void onFailure(GoogleJsonError jsonError, HttpHeaders responseHeaders)
+          throws IOException {
+        GoogleJsonResponseException cause = createJsonResponseException(jsonError, responseHeaders);
+        if (errorExtractor.itemNotFound(cause)) {
+          // Ignore item-not-found errors. We do not have to delete what we cannot find.
+          // This
+          // error typically shows up when we make a request to delete something and the
+          // server
+          // receives the request but we get a retry-able error before we get a response.
+          // During a retry, we no longer find the item because the server had deleted
+          // it already.
+          logger.atFiner().log("Delete folder '%s' not found:%n%s", resourceId, jsonError);
+          successfullDeletionOfFolderResource(
+              resourceId, countOfChildren, folderDeleteBlockingQueue, innerExceptions);
+        } else if (errorExtractor.preconditionNotMet(cause)
+            && attempt <= MAXIMUM_PRECONDITION_FAILURES_IN_DELETE) {
+          logger.atInfo().log(
+              "Precondition not met while deleting '%s'. Attempt %s." + " Retrying:%n%s",
+              resourceId, attempt, jsonError);
+          queueSingleFolderDelete(
+              resourceId,
+              countOfChildren,
+              folderDeleteBlockingQueue,
+              innerExceptions,
+              batchExecutor,
+              attempt + 1);
+        } else {
+          innerExceptions.add(
```

Comment:
This will be logged, correct? Where are we doing this?

Reply:
This is being logged at deleteFolder() function.

Reply:
All the errors that innerException variable gets, those will logged all at once.